packages/cli/src/execution-lifecycle/execution-lifecycle-hooks.ts

```diff
@@ -39,562 +31,450 @@ type HooksSetupParameters = {
-				const { executionId } = this;
-				const { id: workflowId, name: workflowName } = this.workflowData;
-				logger.debug('Executing hook (hookFunctionsPush)', {
-					executionId,
-					pushRef,
-					workflowId,
-				});
-				// Push data to session which started the workflow
-				pushInstance.send(
-					{
-						type: 'executionStarted',
-						data: {
-							executionId,
-							mode: this.mode,
-							startedAt: new Date(),
-							retryOf,
-							workflowId,
-							workflowName,
-							flattedRunData: data?.resultData.runData
-								? stringify(data.resultData.runData)
-								: stringify({}),
-						},
-					},
-					pushRef,
-				);
-			},
-		],
-		workflowExecuteAfter: [
-			async function (this: WorkflowHooks, fullRunData: IRun): Promise<void> {
-				const { executionId } = this;
-				const { id: workflowId } = this.workflowData;
-				logger.debug('Executing hook (hookFunctionsPush)', {
+	hooks.addHandler('nodeExecuteBefore', function (nodeName) {
+		const { executionId } = this;
+		// Push data to session which started workflow before each
+		// node which starts rendering
+		logger.debug(`Executing hook on node "${nodeName}" (hookFunctionsPush)`, {
+			executionId,
+			pushRef,
+			workflowId: this.workflowData.id,
+		});
+
+		pushInstance.send({ type: 'nodeExecuteBefore', data: { executionId, nodeName } }, pushRef);
+	});
+	hooks.addHandler('nodeExecuteAfter', function (nodeName, data) {
+		const { executionId } = this;
+		// Push data to session which started workflow after each rendered node
+		logger.debug(`Executing hook on node "${nodeName}" (hookFunctionsPush)`, {
+			executionId,
+			pushRef,
+			workflowId: this.workflowData.id,
+		});
+
+		pushInstance.send({ type: 'nodeExecuteAfter', data: { executionId, nodeName, data } }, pushRef);
+	});
+	hooks.addHandler('workflowExecuteBefore', function (_workflow, data) {
+		const { executionId } = this;
+		const { id: workflowId, name: workflowName } = this.workflowData;
+		logger.debug('Executing hook (hookFunctionsPush)', {
+			executionId,
+			pushRef,
+			workflowId,
+		});
+		// Push data to session which started the workflow
+		pushInstance.send(
+			{
+				type: 'executionStarted',
+				data: {
 					executionId,
-					pushRef,
+					mode: this.mode,
+					startedAt: new Date(),
+					retryOf,
 					workflowId,
-				});
-
-				const { status } = fullRunData;
-				if (status === 'waiting') {
-					pushInstance.send({ type: 'executionWaiting', data: { executionId } }, pushRef);
-				} else {
-					const rawData = stringify(fullRunData.data);
-					pushInstance.send(
-						{ type: 'executionFinished', data: { executionId, workflowId, status, rawData } },
-						pushRef,
-					);
-				}
+					workflowName,
+					flattedRunData: data?.resultData.runData
+						? stringify(data.resultData.runData)
+						: stringify({}),
+				},
 			},
-		],
-	};
+			pushRef,
+		);
+	});
+	hooks.addHandler('workflowExecuteAfter', function (fullRunData) {
```

Comment:
We could lint to prevent `addHandler` from getting an arrow function callback.

Reply:
array functions are fine here. we only need regular functions when the handler needs access to `this`.